taxcalc/parameters_base.py

```diff
@@ -294,26 +293,75 @@ def _update(self, year_mods):
                          for i in range(0, num_years_to_expand)]
         else:
             inf_rates = None
-        paramvals = self._vals
+        used_names = set()  # set of used parameter names in year_mods dict
         for name, values in year_mods[year].items():
-            # determine inflation indexing status of parameter with name
+            # determine indexing status of parameter with name for year
             if name.endswith('_cpi'):
-                continue
-            if name in paramvals:
-                default_cpi = paramvals[name].get('cpi_inflated', False)
+                continue  # handle elsewhere in this method
+            if name in self._vals:
+                vals_indexed = self._vals[name].get('cpi_inflated', False)
             else:
-                default_cpi = False
-            cpi_inflated = year_mods[year].get(name + '_cpi', default_cpi)
+                msg = 'parameter name {} not in parameter values dictionary'
+                raise ValueError(msg.format(name))
+            name_plus_cpi = name + '_cpi'
+            if name_plus_cpi in year_mods[year].keys():
+                indexed = year_mods[year].get(name_plus_cpi)
+                self._vals[name]['cpi_inflated'] = indexed  # remember status
+                if name_plus_cpi in used_names:
+                    msg = 'parameter {} used twice in year_mods for year {}'
+                    raise ValueError(msg.format(name_plus_cpi, year))
+                else:
+                    used_names.add(name_plus_cpi)
+            else:
+                indexed = vals_indexed
             # set post-reform values of parameter with name
+            if name in used_names:
+                msg = 'parameter {} used twice in year_mods for year {}'
+                raise ValueError(msg.format(name, year))
+            else:
+                used_names.add(name)
             cval = getattr(self, name, None)
             if cval is None:
-                continue
+                msg = 'parameter {} in year_mods for year [] is unknown'
+                raise ValueError(msg.format(name, year))
             nval = self.expand_array(values,
-                                     inflate=cpi_inflated,
+                                     inflate=indexed,
                                      inflation_rates=inf_rates,
                                      num_years=num_years_to_expand)
-            cval[(self.current_year - self.start_year):] = nval
-        self.set_year(self._current_year)
+            cval[(year - self.start_year):] = nval
+        # handle unused parameter names, which all end in _cpi
+        for name, values in year_mods[year].items():
```

Comment:
the logic of this function is now:

```
for all the names in the reform:
    if it ends in '_cpi'
        skip it
    else:
        do the work and keep track of the name

for all the names the names in the form:
    if I haven't seen the name before:
        # these are the names that end in '_cpi'
        do the necessary work
```

So we go through all the names twice. I don't understand why the logic is not:
```
for all the names in the reform:
    if it ends in '_cpi'
        keep track of it and move on
    else:
        do the work and keep track of the name

for all the names I kept track of (which by definition end in '_cpi'):
        do the necessary work
```

That way we don't go through the names twice and avoid extra checking.